### ISSUE TITLE
No warning reported for floating point comparison

### ISSUE THREAD

**Issue by mushenoy** [llvm/llvm-project#83765]
Clang doesnt report any warning for floating point comparision, while GCC reports a warning for the same program.

Sample program:
```
struct number { float no; };

int test (struct number *obj) {
        if (! obj->no ) {
                return 0;
        }
        return 1;
}
```

Compilation with Clang:
No warnings reported
```
# clang -Wfloat-equal -c prog.c
#
```

Compilation with GCC:
```
# gcc -Wfloat-equal -c prog.c
prog.c: In function test:
prog.c:4:13: warning: comparing floating-point with == or != is unsafe [-Wfloat-equal]
    4 |         if (! obj->no ) {
      |             ^
```

Compiler Versions:
```
clang version 16.0.6 
gcc (GCC) 13.2.1
```

**Comment by EugeneZelenko** [comment]
Current `main` produces `-Wfloat-conversion` warning: https://godbolt.org/z/qKK7aoYa9

**Comment by pinskia** [comment]
I am not sure this is a correct warning though:
```
<source>:4:20: warning: implicit conversion turns floating-point number into integer: 'float' to 'bool' [-Wfloat-conversion]
```
It is not turning fp # into a bool but rather `!a` is the same as `a == 0.0`.
well implicit conversions from float to bool is always `a != 0.0` so the warning should be `-Wfloat-equal` rather than `-Wfloat-conversion` here.
